net/gcoap: Confirmable request with piggybacked ACK response

kb2ma opened the pull request:
This PR implements the third step in the confirmable messaging plan in #7192, and depends on #7336. The goal is to use the extended attributes in the request memo to send a confirmable request, including resends on timeout.

Presently, the PR sends the confirmable request using the extended attributes, including a buffer to store the PDU. The next steps are to validate the received response code and type, and then to resend on timeout.

kb2ma commented:
No longer WIP. I plan to rebase ASAP.

kb2ma commented:
Rebased. This PR is a pretty significant milestone for gcoap, and I look forward to getting it merged.

nmeum commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -689,40 +752,88 @@ size_t gcoap_req_send2(const uint8_t *buf, size_t len,
+            if (!_coap_state.resend_bufs[i*GCOAP_PDU_BUF_SIZE]) {
+                memo->msg.data.pdu_buf = &_coap_state.resend_bufs[i*GCOAP_PDU_BUF_SIZE];
+                memcpy(memo->msg.data.pdu_buf, buf, GCOAP_PDU_BUF_SIZE);
+                memo->msg.data.pdu_len = len;
```
You forgot a `break;` here didn't you?

Because if gcoap found a free receive buffer it doesn't need to look for an additional one, right?

kb2ma replied:
Right on, @nmeum! I just pushed your fix to my repository.

smlng commented:
please rebase as #7336 is merged

kb2ma commented:
@smlng, I plan to rebase on #8250. Let me know of any concerns. Thanks.

smlng commented:
please rebase with #8250 merged right now

kb2ma commented:
Rebased; apologies for the delay. Thanks to @miri64 and @bergzand for #8268, which allowed me to run my pyexpect tests on a real board.

kb2ma commented on sys/include/net/gcoap.h:
```diff
@@ -484,6 +493,10 @@ typedef struct {
                                              observe memos */
     gcoap_observe_memo_t observe_memos[GCOAP_OBS_REGISTRATIONS_MAX];
                                         /**< Observed resource registrations */
+    uint8_t resend_bufs[GCOAP_RESEND_BUFS_MAX * GCOAP_PDU_BUF_SIZE];
```
Good idea, makes the intent clearer.

kb2ma commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -91,14 +91,37 @@ static void *_event_loop(void *arg)
+                /* retries remaining; double timeout and resend */
+                else {
+                    /* decrement send limit, and add 1 to advance the timeout */
+                    unsigned i       = COAP_MAX_RETRANSMIT - memo->send_limit-- + 1;
```
I agree. With this update, I removed the 'decrement' comment because it's not really necessary.

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -91,14 +91,37 @@ static void *_event_loop(void *arg)
+                else {
+                    unsigned i       = COAP_MAX_RETRANSMIT - memo->send_limit + 1;
+                    uint32_t timeout = ((uint32_t)COAP_ACK_TIMEOUT << i) * US_PER_SEC;
+                    timeout = random_uint32_range(timeout, timeout * COAP_RANDOM_FACTOR);
```
mhm, I don't like that, bc `COAP_RANDOM_FACTOR` is `1.5` thereby introducing floating point maths here. We know that `timeout` is a multiple of `COAP_ACK_TIMEOUT` (which is two seconds). Hence, I would rather do something like 

```
#define COAP_TIMEOUT_VARIANCE    (500U * US_PER_MS)

uint32_t timeout = ((uint32_t)COAP_ACK_TIMEOUT << i) * US_PER_SEC;
timeout = random_uint32_range(timeout - COAP_TIMEOUT_VARIANCE, timeout + COAP_TIMEOUT_VARIANCE);
```

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -91,14 +91,37 @@ static void *_event_loop(void *arg)
+                }
+                /* retries remaining; double timeout and resend */
+                else {
+                    unsigned i       = COAP_MAX_RETRANSMIT - memo->send_limit + 1;
```
if you put the decrement, i.e. `memo->send_limit--`, above this statement you can omit the `+ 1` at the end of this line.

smlng replied:
I know its likely micro-optimisation, and the compiler will do that already. However, IMHO it makes the code also more readable.

kb2ma replied:
Yes, it's fine and clear with the decrement first. I had it in my head to increase timeout first, but it doesn't matter.

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -91,14 +91,37 @@ static void *_event_loop(void *arg)
+                gcoap_request_memo_t *memo = (gcoap_request_memo_t *)msg_rcvd.content.ptr;
+
+                /* no retries remaining */
+                if (memo->send_limit == GCOAP_SEND_LIMIT_NON
```
I would recommend to enclose each condition in parentheses to clearly separate them.

kb2ma replied:
Yes, I try to do this with new code now.

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -139,9 +167,13 @@ static void _listen(sock_udp_t *sock)
+
+    /* validate class and type for incoming */
+    switch (coap_get_code_class(&pdu)) {
 
```
remove empty line

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -694,42 +756,89 @@ size_t gcoap_req_send2(const uint8_t *buf, size_t len,
-            else {
-                memo->state = GCOAP_MEMO_UNUSED;
-                DEBUG("gcoap: can't wake up mbox; no timeout for msg\n");
+    unsigned msg_type  = (*buf & 0x30) >> 4;
```
why not use `coap_get_type` from nanocoap? Though the variable name indicates what you want here, its not clear why this works, the following should work (I hope):

```
unsigned msg_type = coap_get_type((coap_pkt_t *)buf);
```

smlng replied:
ah, `buf` is not a full packet, hence `coap_get_type` fails, i.e., reads data from where it shouldn't -> segfault.

okay than leave as is (for now).

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -694,42 +756,89 @@ size_t gcoap_req_send2(const uint8_t *buf, size_t len,
+        if (memo->msg.data.pdu_buf) {
+            memo->send_limit = COAP_MAX_RETRANSMIT;
+            timeout = (uint32_t)COAP_ACK_TIMEOUT * US_PER_SEC;
+            timeout = random_uint32_range(timeout, timeout * COAP_RANDOM_FACTOR);
```
same as above, i.e., introduces floating point maths.

smlng commented:
please squash, I'll test in the mean time, so this can go into release 2018.01 as planned.

kb2ma commented:
Thanks! I have squashed on my local workstation. I plan to test and upload later today.

smlng commented:
Looks like you need to rebase (again) to resolve the conflict, too.

kb2ma commented:
Squashed and rebased.

kb2ma commented:
The Murdock build failed after I squashed and rebased. It revealed a type issue in a debug statement on MIPS platforms. It turns out that I was expecting a size_t response from `sock_udp_send()`, but the function really returns an ssize_t. I created a fixup commit so you can review. I can quickly squash this on your approval.

It turns out that there are two other places in gcoap.c that expect `sock_udp_send()` to return a size_t, in `gcoap_req_send2()` and in` gcoap_obs_send()`. I think a fix for the return handling requires a separate PR. Let me know how you'd like to approach it. 

How could this code have been wrong for so long. :-/

smlng commented:
> How could this code have been wrong for so long

well, we recently (yesterday) merged #5166 which changed the handling of debug messages, i.e., previously they were guarded by preprocessor `#if`s and hence not checked if debug wasn't enabled. Now they are checked every time even if debug is disabled.

smlng commented:
> I think a fix for the return handling requires a separate PR

Yes, your minimal fixup seems to be enough to get this PR through Murdock. Anything on top, should be a separate PR.

smlng commented on sys/net/application_layer/gcoap/gcoap.c:
```diff
@@ -91,14 +91,38 @@ static void *_event_loop(void *arg)
+                    uint32_t variance = ((uint32_t)COAP_ACK_VARIANCE << i) * US_PER_SEC;
+                    timeout = random_uint32_range(timeout, timeout + variance);
+
+                    ssize_t res2 = sock_udp_send(&_sock, memo->msg.data.pdu_buf,
```
maybe `bytes` or `bytes_sent` would be a better name than `res2`? Otherwise, looks good, please squash the fixup.

kb2ma commented:
Thanks for the review. Fixed variable name and squashed. Travis and Murdock are happy.

kb2ma commented:
I really appreciate your feedback! Very happy to uncover the bug with handling `sock_udp_send()`. I'll make that my next priority.

smlng commented:
I tested and saw CON and NON request on the wire(shark), looks good to me.
